themes/default/content/blog/managing-github-with-pulumi/index.md

```diff
@@ -0,0 +1,599 @@
+   repos:
+ - role: "read"
+   repos:
+```
+
+This can be read into our Pulumi program as follows:
+
+```go
+// TeamPermissions describes a github team and the levels of permissions available (i.e. admin, maintain)
+type TeamPermissions struct {
+  TeamName    string       `yaml:"team-name"`
+  Permissions []Permission `yaml:"permissions"`
+}
+
+// Permission is a single permission and a list of repos with that permission.
+type Permission struct {
+  Role  string   `yaml:"role"`
+  Repos []string `yaml:"repos"`
+}
+
+#...
+
+func setupTeamRepos(ctx *pulumi.Context, team *github.Team, teamName string) error {
+
+  repoFilePath, err := filepath.Abs("./team-repos/" + teamName + ".yaml")
+  if err != nil {
+     fmt.Println(err.Error())
+     return err
+  }
+  yamlFile, err := ioutil.ReadFile(repoFilePath)
+  if err != nil {
+     return err
+  }
+
+  var teamPerms TeamPermissions
+  err = yaml.Unmarshal(yamlFile, &teamPerms)
+  if err != nil {
+     return err
+  }
+  for _, permission := range teamPerms.Permissions {
+     for _, repo := range permission.Repos {
+        _, err := github.NewTeamRepository(ctx, fmt.Sprintf("%s-%s-%s", repo, teamName, permission.Role), &github.TeamRepositoryArgs{
+           Permission: pulumi.String(permission.Role),
+           Repository: pulumi.String(repo),
+           TeamId:     team.ID(),
+        })
+
+        if err != nil {
+           return err
+        }
+     }
+  }
+  return nil
+}
+```
+
+And the output of running `pulumi up`:
+
+```bash
+Updating (prod)
+
+     Type                            Name                      Status
+     pulumi:pulumi:Stack             team-mgmt-prod
+ +   ├─ github:index:TeamRepository  owls-animals-admin        created
+ +   ├─ github:index:TeamRepository  capybaras-animals-admin   created
+
+
+Resources:
+    + 2 created
+    2 changes. 4 unchanged
+```
+
+With this addition, we can now combine teams with zero access disruptions as follows:
+
+1. Create the new team in the org config
+2. Create a TeamRepository config for the new team with desired permission levels
+3. Open a pull request and let CI and Pulumi do the rest.
+
+Now we can develop standards around repository permissions based on teams’ roles and areas of responsibility, and we have tooling in place that can maintain these standards for everyone in the org to see. In fact, we can standardize the meaning of “code ownership” via access levels in this way - one centralized management location rather than asking your grandboss to dig through the UI for you.
+
+## Future challenges
+
+We want to automate some of these steps: adding new teammates to the org should also automatically add them to this repository so they can self-serve their onboarding and see Pulumi work in action!
+
+To prevent state drift, we will run a regular reconciliation job in CI.
+
+Of course, we will add other Pulumi providers on top of this org structure so we can sync our Slack teams, our Bamboo teams, and our gsuite groups.
```

Comment:
This is more of a thought. At other companies, I'd been instructed not to share what platforms we use for security reasons (you can't hack what you don't know is there). Thoughts?

```suggestion
Of course, we will add other Pulumi providers on top of this org structure so we can sync our other tooling.
```

Reply:
Honestly, security reasons aside, I'm not sure readers will care what other tools we have here - your wording is just much more useful overall.